### DISCUSSION TITLE
Midi Library instance as member variable

### DISCUSSION THREAD

**Discussion by ndsizeif** [FortySevenEffects/arduino_midi_library#221] [Q&A]
I have created a class for hardware devices and I would like to be able to assign
a specific Midi interface for each.  Before setup in the main sketch I create 4 instances of the Midi library
to be used with 4 different device objects.

```
MIDI_CREATE_INSTANCE(SoftwareSerial, softSerial1, MIDIA);  // device A
MIDI_CREATE_INSTANCE(SoftwareSerial, softSerial2, MIDIB);  // device B
MIDI_CREATE_INSTANCE(SoftwareSerial, softSerial3, MIDIC);  // device C
MIDI_CREATE_INSTANCE(SoftwareSerial, softSerial4, MIDID);  // device D
```

In my class document I don't know the correct syntax to use in order to  create a constructor than allows me to assign 
one of the MIDI instances to the device object.

```
#include <MIDI.h>
#include "Arduino.h"
 
class HardwareDevice
{  
  public:
 
    String Name;
    int Channel;
    ???? Interface;
  
  HardwareDevice(String name, int channel, ???? interface) {
    this->Name = name;
    this->Channel = channel;
    this->Interface = interface;
 }
```

I did see a similar question in [this issue thread](https://github.com/FortySevenEffects/arduino_midi_library/issues/165), but the use case was
different enough to lead me off course.  If anyone could point me in the right direction again,
I would appreciate it.

**Comment by lathoub** [comment]
Why create a wrapper class `HardwareDevice`?

```cpp
MIDI_CREATE_INSTANCE(SoftwareSerial, softSerial1, MIDIA);  // device A
MIDI_CREATE_INSTANCE(SoftwareSerial, softSerial2, MIDIB);  // device B
MIDI_CREATE_INSTANCE(SoftwareSerial, softSerial3, MIDIC);  // device C
MIDI_CREATE_INSTANCE(SoftwareSerial, softSerial4, MIDID);  // device D

void setup() {
  ...
  MIDIA.begin(MIDI_CHANNEL_OMNI); // read from all channels
  MIDIB.begin(1); // read from channel 1
  MIDIC.begin(2); // read from channel 2
  MIDID.begin(3); // read from channel 3
  ...
}

void loop() {
  ...
  MIDIA.read()
  MIDIB.read()
  MIDIC.read()
  MIDID.read()

  MIDIA.sendNoteOff(42, 0, 1);
}

```

**Reply by ndsizeif** [reply]
> Why create a wrapper class `HardwareDevice`?

Thank you for the reply.

My HardwareDevice class has a lot of functions which aren't midi-related at all, so I didn't include them in the example code.  
Each HardwareDevice object can have other objects like leds/storage/displays assigned to it.  In the way I can initialize those objects and then assign them to the device of my choosing, I was hoping to do the same with the Midi library.  I just don't know the syntax for this.

My class doesn't use read() or sendNote() functions.  I mainly use it for program change and system exclusive messages.

A rudimentary example using LED objects:

```
class hardwareDevice 
{
  String Name
  char Letter
  int LED
  int Channel
  String Description
  ???? MidiInterface

  hardwareDevice(String name, char letter, int led, int channel, String description, ???? interface) {
  
    this->Name = name;
    this->Letter = letter;
    this->LED = led;
    this->Channel = channel;
    this->Description = description;
    this->MidiInterface = interface;
  }

  void genericBlink() {
    digitalWrite(LED, HIGH);
    delay(500);
    digitalWrite(LED, LOW);
  }
  void funkyBlink() {
     digitalWrite(LED, HIGH);
     delay(100);
     digitalWrite(LED, LOW);
     delay(1500);
   }
  void changeChannel() {
    if (LED == HIGH) {
      Channel = 16:
    }
   else
      Channel = 1;  
   }
};

#define RLED 2
#define GLED 3
#define BLED 4

hardwareDevice Red = hardwareDevice("Red Machine, 'A', RLED, 2,  "This one is red and uses midi channel 2");
hardwareDevice Green = hardwareDevice("Green Box", "B", GLED, 4, "This one is green and uses midi channel 4");
hardwareDevice Blue = hardwareDevice("Blue Bureau", "C", BLED, 16, "This one is blue and uses midi channel 16");

setup() { 

  pinMode(RLED, OUTPUT);
  pinMode(GLED, OUTPUT);
  pinMode(BLED, OUTPUT);

}

loop() {

  Green.genericBlink();
  Blue.funkyBlink();
  Blue.changeChannel();
}
```
Thanks again for your help.

**Reply by lathoub** [reply]
you can create an instance in the `hardwareDevice` class and initialise in the constructor.
 https://github.com/FortySevenEffects/arduino_midi_library/blob/d501f4bb3dfff548a7cef4d85371bec0b438bdd7/src/serialMIDI.h#L129-L130

**Reply by ndsizeif** [reply]
Can you provide an example of the syntax for this?  

I've tried many different ways to get the results i want.  Using the way outlined in this thread,
currently I have a member function that calls MIDI_CREATE_INSTANCE()
but it requires 3 arguments and I only know the syntax to provide 2.
